integration/hammer.go

```diff
@@ -0,0 +1,366 @@
+// out across the MMD period.
+type pendingCerts [sctCount]*submittedCert
+
+func (pc *pendingCerts) canAppend(now time.Time, mmd time.Duration) bool {
+	if pc[sctCount-1] != nil {
+		return false // full already
+	}
+	if pc[0] == nil {
+		return true // nothing yet
+	}
+	// Only allow append if enough time has passed, namely MMD/#savedSCTs.
+	last := sctCount - 1
+	for ; last >= 0; last-- {
+		if pc[last] != nil {
+			break
+		}
+	}
+	lastTime := TimeFromMS(pc[last].sct.Timestamp)
+	nextTime := lastTime.Add(mmd / sctCount)
+	return now.After(nextTime)
+}
+
+func (pc *pendingCerts) appendCert(submitted *submittedCert) {
+	// Require caller to have checked canAppend().
+	which := 0
+	for ; which < sctCount; which++ {
+		if pc[which] == nil {
+			break
+		}
+	}
+	pc[which] = submitted
+}
+
+// popIfMMDPassed returns the oldest submitted certificate (and removes it) if the
+// maximum merge delay has passed, i.e. it is expected to be integrated as of now.
+func (pc *pendingCerts) popIfMMDPassed(now time.Time) *submittedCert {
+	if pc[0] == nil {
+		return nil
+	}
+	submitted := pc[0]
+	if !now.After(submitted.integrateBy) {
+		// Oldest cert not due to be integrated yet, so neither will any others.
+		return nil
+	}
+	// Can pop the oldest cert and shuffle the others along, which make room for
+	// another cert to be stored.
+	for i := 0; i < (sctCount - 1); i++ {
+		pc[i] = pc[i+1]
+	}
+	pc[sctCount-1] = nil
+	return submitted
+}
+
+// HammerCTLog performs load/stress operations according to given config.
+func HammerCTLog(cfg HammerConfig) error {
+	opts := jsonclient.Options{}
+	if cfg.LogCfg.PubKeyPEMFile != "" {
+		pubkey, err := ioutil.ReadFile(cfg.LogCfg.PubKeyPEMFile)
+		if err != nil {
+			return fmt.Errorf("failed to get public key contents: %v", err)
+		}
+		opts.PublicKey = string(pubkey)
+	}
+	var pool ClientPool
+	for _, s := range strings.Split(cfg.Servers, ",") {
+		c, err := client.New("http://"+s+"/"+cfg.LogCfg.Prefix, nil, opts)
+		if err != nil {
+			return fmt.Errorf("failed to create LogClient instance: %v", err)
+		}
+		pool = append(pool, c)
+	}
+	ctx := context.Background()
+
+	stats := newWantStats(cfg.LogCfg.LogID)
+	// STHs are arranged from later to earlier (so [0] is the most recent), and the
+	// discovery of new STHs will push older ones off the end.
+	sth := [sthCount]*ct.SignedTreeHead{}
+	// Submitted certs also run from later to earlier, but the discovery of new SCTs
+	// does not affect the existing contents of the array, so if the array is full it
+	// keeps the same elements.  Instead, the oldest entry is removed (and a space
+	// created) when we are able to get an inclusion proof for it.
+	var pending pendingCerts
+
+	for count := uint64(1); count < cfg.Operations; count++ {
+		ep := cfg.EPBias.Choose()
+		glog.V(3).Infof("perform %s operation", ep)
+		status := http.StatusOK
+		switch ep {
+		case ctfe.AddChainName:
+			chain, err := MakeCertChain(cfg.LeafChain, cfg.LeafCert, cfg.CACert, cfg.Signer)
```

Comment:
Consider moving these cases into sub-functions to reduce the complexity and size of the outer function. 
Could also make these sub-cases reusable.